reconcile/ocm_machine_pools.py

```diff
@@ -1,137 +1,332 @@
+                        subnet=machine_pool.get("subnet"),
+                        cluster=cluster.name,
+                    )
+                )
+        else:
+            machine_pools = ocm.get_machine_pools(cluster.name)
+            for machine_pool in machine_pools:
+                current_state[cluster.name]["machine_pools"].append(
+                    MachinePool(
+                        id=machine_pool["id"],
+                        replicas=machine_pool["replicas"],
+                        instance_type=machine_pool["instance_type"],
+                        taints=machine_pool.get("taints"),
+                        labels=machine_pool.get("labels"),
+                        cluster=cluster.name,
+                    )
+                )
 
-    return ocm_map, current_state
+    return current_state
 
 
-def fetch_desired_state(clusters):
-    desired_state = []
+def fetch_desired_state(
+    clusters: Iterable[ClusterV1],
+):
+    desired_state = {}
     for cluster in clusters:
-        cluster_name = cluster["name"]
-        machine_pools = cluster["machinePools"]
-        for machine_pool in machine_pools:
-            machine_pool["cluster"] = cluster_name
-            labels = machine_pool.pop("labels")
-            if labels:
-                machine_pool["labels"] = json.loads(labels)
-            taints = machine_pool.pop("taints")
-            if taints:
-                machine_pool["taints"] = taints
-            desired_state.append(machine_pool)
+        if cluster.machine_pools:
+            desired_state[cluster.name] = {
+                "machine_pools": cluster.machine_pools,
+                "hypershift": cluster.spec.hypershift,
+            }
 
     return desired_state
 
 
-def calculate_diff(current_state, desired_state):
-    diffs = []
-    err = False
-    for d in desired_state:
-        c = [
-            c
-            for c in current_state
-            if d["cluster"] == c["cluster"] and d["id"] == c["id"]
-        ]
-        if not c:
-            d["action"] = "create"
-            diffs.append(d)
-            continue
-        if len(c) != 1:
-            logging.error(f"duplicate id found in {d['cluster']}")
-            err = True
-            continue
-        c = c[0]
-        if c == d:
-            continue
-        if d["instance_type"] != c["instance_type"]:
-            logging.error(
-                f"can not update instance type for existing "
-                f"machine pool {d['id']} in {d['cluster']}"
-            )
-            err = True
-            continue
-        d.pop("instance_type")
-        for key in ["labels", "taints"]:
-            if c.get(key, None) != d.get(key, None):
-                # https://github.com/openshift/machine-api-operator/blob/master/FAQ.md
-                logging.warning(
-                    f"update {key} for machine pool {d['id']} "
-                    f"will only be applied to new Nodes"
-                )
```

Comment:
is this warning worth bringing over to the new code?

Reply:
added